### PR TITLE
@types/koa-joi-router: update version and fix some issues

### PR DESCRIPTION
**Pull request by hirochachacha**
* extend Koa.Request rather than defining its own type.
* support complex handler types.
* support multiple method.
* add use, param, head, option, get, post, put, patch, delete

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [koajs/joi-router](https://github.com/koajs/joi-router)
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/koa-joi-router/index.d.ts [hunk 1]
```diff
@@ -31,20 +46,22 @@ declare namespace createRouter {
...
-    interface Context extends Koa.Context {
-        request: Request;
-    }
-
     interface Router {
         routes: Spec[];
         route(spec: Spec|Spec[]): Router;
         middleware(): Koa.Middleware;
-        prefix(path: string): void;
+        prefix(path: string): Router;
```
Comment (wingsbob) [inline types/koa-joi-router/index.d.ts]: `prefix` doesn't return `this`, it's delegated to the koa-router, which returns that router.

Same for `use` and `param`

https://github.com/koajs/joi-router/blob/master/joi-router.js#L42-L45

Comment (hirochachacha) [reply types/koa-joi-router/index.d.ts]: Thank you for your review. I decided to reuse the koa-router type.

**Review by wingsbob — APPROVED** [review]
lgtm

I'm guessing `interface NestedHandler extends ReadonlyArray<PartialHandler|FullHandler|NestedHandler> {}` can't be `type NestedHandler 
 = ReadonlyArray<PartialHandler|FullHandler|NestedHandler>;` because it's self-referencing? Haven't had a chance to check if it'd compile, and looks fine otherwise 😄

**Comment by hirochachacha** [comment]
> I'm guessing interface NestedHandler extends ReadonlyArray<PartialHandler|FullHandler|NestedHandler> {} can't be type NestedHandler = ReadonlyArray<PartialHandler|FullHandler|NestedHandler>; because it's self-referencing?

You're right. It looks like TypeScript has a limitation in recursive type definitions.
```
$tsc

index.d.ts:28:10 - error TS2456: Type alias 'NestedHandler' circularly references itself.

28     type NestedHandler = ReadonlyArray<PartialHandler|FullHandler|NestedHandler>;
            ~~~~~~~~~~~~~
```
That is a work-around I've learned yesterday. https://github.com/Microsoft/TypeScript/issues/3496#issuecomment-128553540

**Comment by PranavSenthilnathan** [comment]
Congratulations on your first DefinitelyTyped contribution! Thank you for being a part of the community!

**Pull request merged**